vertical-pod-autoscaler/pkg/admission-controller/logic/recommendation_provider.go

```diff
@@ -61,17 +63,35 @@ func NewRecommendationProvider(vpaLister vpa_lister.VerticalPodAutoscalerLister,
 	}
 }
 
-func limitMatchesRequest(limit, request *resource.Quantity) bool {
-	// Limit not set, it cannot match the request.
-	if limit == nil || limit.Value() == 0 {
-		return false
+func getProportionalLimit(originalLimit, originalRequest, recommendedRequest *resource.Quantity) *resource.Quantity {
+	// originalLimit not set, don't set limit.
+	if originalLimit == nil || originalLimit.Value() == 0 {
+		return nil
+	}
+	// originalLimit set but originalRequest not set - K8s will treat the pod as if they were equal,
+	// recommend limit equal to request
+	if originalRequest == nil || originalRequest.Value() == 0 {
+		return recommendedRequest
 	}
-	// Limit set but request not set - K8s will treat the pod as if they were equal.
-	if request == nil || request.Value() == 0 {
-		return true
+	// originalLimit and originalRequest are set. If they are equal recommend limit equal to request.
+	if originalRequest.MilliValue() == originalLimit.MilliValue() {
+		return recommendedRequest
 	}
-	// Limit and request are set. Check if they are equal.
-	return request.Value() == limit.Value()
+
+	// Input and output milli values should fit in int64 but intermediate values might be bigger.
+	originalMilliRequest := big.NewInt(originalRequest.MilliValue())
+	originalMilliLimit := big.NewInt(originalLimit.MilliValue())
+	recommendedMilliRequest := big.NewInt(recommendedRequest.MilliValue())
+	var recommendedMilliLimit big.Int
+	recommendedMilliLimit.Mul(recommendedMilliRequest, originalMilliLimit)
+	recommendedMilliLimit.Div(&recommendedMilliLimit, originalMilliRequest)
+	if recommendedMilliLimit.IsInt64() {
+		return resource.NewMilliQuantity(recommendedMilliLimit.Int64(), recommendedRequest.Format)
+	}
+	klog.Warningf(
+		"Trying to keep proportion between original millilimit of %s and original millirequest of %s with recommended millirequest of %s the recommended millilimit of %s doesn't fit in int64, capping to MaxInt64",
```

Comment:
nit: the milli... makes this really long and convoluted. How about:
Failed to keep limit to request proportion of %s to %s with recommended request of %s milliUnits; doesn't fit in int64. Capping limit to MaxInt64

Reply:
Done.